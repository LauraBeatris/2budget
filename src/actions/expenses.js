import uuid from "uuid";

//ADD_EXPENSE
export const addExpense = ({
  description = "",
  note = "",
  amount = 0,
  createdAt = 0
} = {}) => {
  return {
    type: "ADD_EXPENSE",
    expense: {
      // Setting a id
      id: uuid(),
      description,
      note,
      amount,
      createdAt
    }
  };
};

//REMOVE_EXPENSE
export const removeExpense = ({ id } = {}) => ({
  type: "REMOVE_EXPENSE",
  id
});

//EDIT_EXPENSE
export const editExpense = (id, update) => ({
  type: "EDIT_EXPENSE",
  id,
  update
});
