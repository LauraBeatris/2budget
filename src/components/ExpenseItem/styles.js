import styled from "styled-components";

export const ItemContainer = styled.div`
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  width: 30%;
  font-size: 2rem;
  padding: 1rem 1rem;
  border-bottom: 1px solid rgb(77, 81, 94);
  transition: all 0.25s;

  &:hover {
    transform: scale(1.05);
  }

  a.description {
    text-decoration: none;
    font-family: "Montserrat", sans-serif;
    color: #6dbfcc;
    font-weight: 600;
  }

  p {
    color: rgb(77, 81, 94);
    font-weight: 500;
  }
`;
