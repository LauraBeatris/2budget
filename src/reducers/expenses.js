const expensesReducerDefaultState = [];

const expensesReducer = (state = expensesReducerDefaultState, action) => {
  switch (action.type) {
    case "ADD_EXPENSE":
      return [...state, action.expense];
    case "REMOVE_EXPENSE":
      return state.filter(({ id }) => id !== action.id);
    case "EDIT_EXPENSE":
      return state.map(expense => {
        if (expense.id === action.id) {
          return {
            ...expense,
            ...action.update // overriden the expense items with the action updates icons
          };
        } else {
          return expense;
        }
      });
    default:
      return state;
  }
};

export default expensesReducer;
