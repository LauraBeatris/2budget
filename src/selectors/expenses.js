import moment from "moment";

// Get visible expenses
export const getVisibleExpenses = (
  expenses,
  { text, sortBy, startDate, endDate }
) => {
  return expenses
    .filter(expense => {
      const createdAtMoment = moment(expense.createdAt);
      const startDateMatch = startDate
        ? startDate.isSameOrBefore(createdAtMoment, "day")
        : true;
      const endDateMatch = endDate
        ? endDate.isSameOrAfter(createdAtMoment, "day")
        : true;

      const textMatch = expense.description
        .toLowerCase()
        .includes(text.toLowerCase());

      return startDateMatch && endDateMatch && textMatch;
    })
    .sort((a, b) => {
      if (sortBy === "date") {
        return b.createdAt < a.createdAt ? -1 : 1;
      } else if (sortBy === "amount") {
        return a.amount > b.amount ? -1 : 1;
      } else {
        return null;
      }
    });
};

export const getExpenseTotal = expenses => {
  if (expenses.length === 0) {
    return 0;
  } else if (expenses.length > 1) {
    const amounts = expenses.map(expense => expense.amount);
    return amounts.reduce((sum, value) => sum + value, 1);
  } else {
    return expenses[0].amount;
  }
};
