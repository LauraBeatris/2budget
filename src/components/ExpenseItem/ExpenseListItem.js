import React from "react";
import { ItemContainer } from "./styles";
import { Link } from "react-router-dom";
import moment from "moment";

// Removing an expense, dispatching the remove expense action
// Every connected element receives the dispatch as a prop

const ExpenseItem = ({ id, description, amount, createdAt }) => (
  <ItemContainer>
    <Link className="description" to={`/edit/${id}`}>
      {description}
    </Link>
    <p className="amount">${amount}</p>
    <p className="createdAt">{moment(createdAt).format("MM/DD/YYYY")}</p>
  </ItemContainer>
);

export default ExpenseItem;
