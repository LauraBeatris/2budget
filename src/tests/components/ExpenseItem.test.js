import React from "react";
import { shallow } from "enzyme";
import { ExpenseItem } from "../../components/ExpenseItem/ExpenseListItem";
import expenses from "../fixtures/expenses";

test("should render ExpenseItem with the expense data", () => {
  const wrapper = shallow(<ExpenseItem {...expenses[0]} />);
  expect(wrapper).toMatchSnapshot();
});
