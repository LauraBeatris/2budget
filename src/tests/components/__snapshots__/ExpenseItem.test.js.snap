// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseItem with the expense data 1`] = `
<.sc-bdVaJa>
  <Link
    className="description"
    to="/edit/1"
  >
    Water rent
  </Link>
  <p
    className="amount"
  >
    $
    195
  </p>
  <p
    className="createdAt"
  >
    31/12/1969
  </p>
</.sc-bdVaJa>
`;
